### PR TITLE
feat: Add unit tests

### PR DESCRIPTION
**Pull request by Cucumberrbob**
Writing this latest edit made me realise I want to do a little more work on this. It's ready for review, but imo not ready to merge

I've been working on adding unit tests.

Since this is the *introduction* of unit tests, **I'd like to ask any reviewers to be *really* nitpicky**. Other contributors may use these as templates for future unit tests, so it's important this pr exemplifies ideal tests. Stuff like names (`Mocks`, `Fake*`, the data used in tests, `MemberData` classes, etc), and patterns for the unit tests (asserting log lines, where the fake/fixture data comes from) are really important to get right now imo.

My minimal goals for this PR (figuring out the *how* of adding unit tests) is mostly completed. I think I've presented solutions for pretty much all of the dependency injection problems we're likely to have.  Again, super nitpicky feedback is really useful here, as I'll be implementing similar patterns for other modules (e.g. the `AllDebridNetClientFactory` equivalent for `RealDebrid`, `Premiumize`, etc).
I've also tried to be minimalist in my changes: for example in `Torrents`, I've added an `IFileSystem fileSystem` to the constructor, but I've only used it in the method I've written tests for.

## Tests

- `Torrents#RunTorrentComplete`
  Check it runs when we expect (command set/not set), and logs as we expect (stdout/stderr)
- `DownloadHelper`
  Basically full coverage - the only lines that are not covered are unreachable
  Also documented a slightly weird edge-case where when `download.FileName` is null and `download.Link` is a url without a path, `GetDownloadPath` returns the `torrentPath` (a path to a directory), not a path to a file.
- `DownloadStationDownloader`
  These changes are my best so far imo. In large part because the class has been written in a way friendly to unit testing, and the `Synology.Api.Client` exposes interfaces we can mock.
  Also added a `DelayProvider` so we don't have to actually wait for `Task.Delay` in the tests.
- `AllDebridTorrentClient`
  Following on from rogerfar/AllDebrid.NET#9, I separated `GetClient` out from `AllDebridTorrentClient`, into `AllDebridNetClientFactory`. We can then easily mock the `ClientFactory` in the tests.
  I've tested mapping torrents from `AD` API responses to `rdt-client` `TorrenClientTorrentt`s (via `GetTorrent`), `GetTorrent` itself (both partial and full sync), and file filtering (include/exclude regex, DownloadMinSize) in `GetDownloadLinks`. I think there is some room for improvement: the `include/exclude` filenames/structure could be expanded, and we should almost certainly test the `GetSymlinkPath` logic given the headache it's caused in the past. However, I want to hold off on this until I head back from `AD` about whether `/api/v4.1/magnets/status` with an `id` specified should give back file information or not (their docs say it doesn't but it does), as this is where `rdt-client` currently gets the file structure from.

## Notable changes
- Introduction of interfaces for `Service.Downloads` and `Data.TorrentData` - this lets us mock them with `Moq`
- Introduction of interfaces `IProcess` and  `IProcessFactory`- lets us mock `Process`
- Introduction of factory for getting an `AllDebrid` api client
- Introduction of `IDelayProvider` so we don't have to actually wait for `Task.Delay` in the tests
  - Could the argument be `delayProvider.Delay(Int64 retryCount)`? rather than `milliseconds`?
- Added/updated stuff in `DiConfig`s

## CI
I've also got a branch I'll try to keep rebased onto `cucumberrbob/feat/unit-tests` with a GitHub Action to `dotnet test`. Here's a [run](https://github.com/Cucumberrbob/rdt-client/actions/runs/13379636784/job/37365795484) from [1d15b0d](https://github.com/rogerfar/rdt-client/pull/691/commits/1d15b0d6516f309f52cc0811b1c8065ba30b3911).

## Patterns I've attempted to follow 
I'm still a csharp noob, so I may well have misinterpreted some of these patterns, but here's a rough outline of the patterns I've tried to follow:
- Arrange/Act/Assert  (structure of tests)
  Not sure if the comments are a thing, or if it's just I've been reading a lot of example code, but this seems to be common. The pattern itself of separating three steps is absolutely a thing though.
  Is `var exception = Assert.ThrowsAsync<Exception>(myFunc)` really an `Act` phase? There's an assertion right there.. But I'm not sure how else to do this kinda thing; try/catching feels worse
- MethodName/Scenario/Behaviour (naming of tests)
  I think I'v e done decently well with this (feedback is super helpful if I'm wrong here)
  
But I'v used a bunch of magic strings and repeated my self a bunch across a variety of tests. **This is something I want feedback on especially**; I'm not entirely sure how to refactor these out given the context they're used in.

[Microsoft Testing Best Practices ](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices)

### CODE REVIEW

**Comment by Cucumberrbob** [comment]
Hey @rogerfar @asylumexp @Sculas, @Wald764 @YvesPa  (pinging everyone who's committed in the last 3 months)
Would you be able to review this PR? I'd really appreciate an in depth review, because it's going into new territory - so I think we need to get it right the first time. Any feedback is welcome, you do not have to review the entire thing. 
I completely understand if you don't have the time/energy though! It's almost 2000 lines, so it's not exactly small.

File: server/RdtClient.Service.Test/Services/TorrentsTest.cs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    [Theory]
+    [MemberData(nameof(TorrentAndDownload))]
+    public async Task RunTorrentComplete_WhenStdErr_Logs(Torrent torrent, List<Download> downloads)
+    {
+        // Arrange
+        var settings = new DbSettings()
+        {
+            General = new DbSettingsGeneral()
+            {
+                RunOnTorrentCompleteFileName = "/bin/echo"
```
Comment (Sculas) [inline server/RdtClient.Service.Test/Services/TorrentsTest.cs:252]: This probably causes the test to fail on Windows. Is there a platform-agnostic variant you can use for this?

Changing this to an if-statement that uses `C:\Windows\System32\cmd.exe` on Windows and `/bin/echo` on Unix, for example, is also an option if it doesn't matter what you need to run here. If you need the output, just add `/K echo` as an argument on Windows.

There are multiple places where this happens, but I'll only note this one.

**Review by Sculas — APPROVED** [review]
It looks good to me, except for this small thing I noticed. I don't have very much time, so my review is not 100% thorough, but I did my best to check everything.

File: server/RdtClient.Service.Test/Services/TorrentsTest.cs (hunk 1 above)
Comment (Cucumberrbob) [reply server/RdtClient.Service.Test/Services/TorrentsTest.cs:252]: The process isn’t actually getting run, it’s all mocked (that’s why I added `IProcess` and `IProcessFactory`).
It might be worth adding a comment somewhere explaining this if it’s not immediately obvious.

Thank you so much for taking the time to review!

Comment (rogerfar) [reply server/RdtClient.Service.Test/Services/TorrentsTest.cs:252]: No mocking Process is fine, you want to test the unit, not do an e2e test at this point. We also don't know what the other end looks like. We just need to know what the process gets called with the correct parameters as documented.

**Review by rogerfar — APPROVED** [review]
Your tests look good! I'm more a fan of Xunit, worth looking into if you're interested.

**Comment by rogerfar** [comment]
I fixed few formatting issues that popup in the output, but also this shows:

The member referenced by the MemberData attribute returns untyped data rows, such as object[]. Consider using TheoryData<> as the return type to provide better type safety.

Just FYI.

**Comment by rogerfar** [comment]
Also, it's worth using a coverage tool like dotCover to check how much of the project is unit tested. Don't take it as a goal to reach 100% coverage, but it's good to see missing edge cases.

**Comment by Cucumberrbob** [comment]
> Your tests look good! I'm more a fan of Xunit, worth looking into if you're interested.

Thank you!
I thought I was using xunit though?
https://github.com/rogerfar/rdt-client/pull/691/files#diff-e0686c524385fefbad56cc6371b7171ee9a928725a3629d954548c81eeb1f03aR19-R20

> Consider using TheoryData<> as the return type to provide better type safety.

That would be perfect, I knew the ;`object[]` thing felt weird, but almost all the stack overflow posts etc I saw used this

> it's worth using a coverage tool like dotCover to check how much of the project is unit tested

Yep I've got this in Rider. It's useful to be able to see the paths the tests go down, but I won't use test coverage % as a goal.

**Comment by rogerfar** [comment]
Sorry yes I just noticed, I saw the `[Theory]` and no explicit top level includes, so assumed it was another framework.
